include/wil/registry.h

```diff
@@ -782,12 +782,13 @@ namespace wil
          * \return HRESULT error code indicating success or failure (does not throw C++ exceptions)
          */
         template <>
-        inline HRESULT set_value_nothrow(HKEY key, _In_opt_ PCWSTR subkey, PCWSTR value_name, const ::std::vector<::std::wstring>& data) WI_NOEXCEPT
+        inline HRESULT set_value_nothrow(HKEY key, _In_opt_ PCWSTR subkey, PCWSTR value_name, const ::std::vector<::std::wstring>& data) WI_NOEXCEPT try
         {
             const auto multiStringWcharVector(reg_view_details::get_multistring_from_wstrings(::std::begin(data), ::std::end(data)));
             const reg_view_details::reg_view_nothrow regview{ key };
             return regview.set_value(subkey, value_name, multiStringWcharVector, REG_MULTI_SZ);
         }
+        CATCH_RETURN();

```

Comment:
reg_view_details::get_multistring_from_wstrings can throw. this function is only available if the user has already included the STL headers for vector and string, so exceptions must be enabled. (it's also within #if defined(WIL_ENABLE_EXCEPTIONS))

Reply:
I suggest we just don't offer this `get_value_nothrow` specialization then.

The value of `_nothrow` functions is that they can be used in code with exceptions disabled. If exceptions are enabled, callers should call the exceptional versions and `CATCH_RETURN` on *their* side if needed.

I know we have plenty of code with exceptions enabled that needs to return HRESULTs, but when I see a WIL `_nothrow` function, I expect that there's no exceptions in the code whatsoever. Doing a behind-the-scenes catch is not what I expect.

We could have this specific specialization fail with a helpful compiler error, but I don't like the idea that we conditionally include this nothrow variant only when exceptions are enabled *and* this variant secretly can throw (and just catches the exception).

Reply:
This has come up many times: I will continue to argue against the assumption that "The value of _nothrow functions is that they can be used in code with exceptions disabled." The exception specification of a function is tied to the interface contract with the caller - quite often interfaces will want to declare noexcept on functions and methods that use types that would otherwise throw.

_nothrow becomes particularly useful when it's being called in a function that would otherwise have to tightly wrap a try/catch around a single function call; that is a bad interface pattern if it can't be easily avoided.

If the WIL owners will publicly document & dictation, "All WIL _nothrow functions exist for when exceptions are completely disabled" then sure - I'll rip up this code a bit so absolutely no STL is accessible in any _nothrow functions.